src/rgw/rgw_putobj_processor.cc

```diff
@@ -457,4 +457,161 @@ int MultipartObjectProcessor::complete(size_t accounted_size,
   return 0;
 }
 
+int AppendObjectProcessor::process_first_chunk(bufferlist &&data, rgw::putobj::DataProcessor **processor)
+{
+  int r = writer.write_exclusive(data);
+  if (r < 0) {
+    return r;
+  }
+  *processor = &stripe;
+  return 0;
+}
+
+int AppendObjectProcessor::prepare()
+{
+  RGWObjState *astate;
+  int r = store->get_obj_state(static_cast<RGWObjectCtx *>(s->obj_ctx), bucket_info, rgw_obj(s->bucket, s->object), &astate);
+  if (r < 0) {
+    return r;
+  }
+  cur_size = astate->size;
+  *cur_accounted_size = astate->accounted_size;
+  if (!astate->exists) {
+    if (position != 0) {
+      ldout(store->ctx(), 5) << "ERROR: Append position should be zero" << dendl;
+      return -ERR_POSITION_NOT_EQUAL_TO_LENGTH;
+    } else {
+      cur_part_num = 1;
+      //set the prefix
+      char buf[33];
+      gen_rand_alphanumeric(store->ctx(), buf, sizeof(buf) - 1);
```

Comment:
okay, so two racing appends on a nonexistent object will write to separate tail objects because of the random prefix, and the race is resolved on atomic complete(). this also covers a race between an append and normal put

and two racing appends to an existing object will use the same prefix and part number, so the write_exclusive() in process_first_chunk() will only succeed for the first writer

am i understanding this right?

Reply:
Yes~